### ISSUE TITLE
TLS with custom cert

### ISSUE THREAD

**Issue by ben** [gliderlabs/logspout#147]
Logspout works great for shipping logs to Papertrail. Except if you want to use TLS, which (for papertrail at least) requires a [custom cert](http://help.papertrailapp.com/kb/configuration/encrypting-remote-syslog-with-tls-ssl/).

My ideal solution for this would be to volume-map in the cert I want to trust to a known location (like `/tmp/cert.pem`), and Logspout would load the cert if it existed, and use that for TLS to Papertrail instead of the system certs.

**Comment by dan-compton** [comment]
@ben afaik, crypto/tls uses the hosts root ca set if RootCAs is nil so this might already be possible with an exec to update the root ca set

**Comment by iiro** [comment]
Hi @ben,

did you find a solution for this?

**Comment by troy** [comment]
FYI, Papertrail doesn't generally require a custom cert[1]. The file linked to above is a root CA and intermediate certs. Whether or not it's required depends on the OS root CA bundle (and whether the connecting app chooses to use that OS-supplied bundle, if available). The Papertrail cert file is provided because:

1. many root bundles are broken or incomplete ([a few examples](https://www.google.com/search?q=openssl+broken+root+bundle)). Also, a few log senders/libraries initialize the TLS library in a way that intentionally disables the OS-provided bundles. It's hard for a user to know that their system or sender has this problem without spending a lot of time troubleshooting.
2. depending on the sending app, Papertrail destinations created in 2012 and earlier may require an intermediate cert. For those pre-2013 destinations, Papertrail doesn't present a complete CA chain, though only some apps/SSL libraries care.

With Papertrail log destinations created in the last 2-3 years (and connections made from modern systems and clients -- that is, those honoring a typical set of root CAs), it's possible to use the system root CA bundle.

[1]: for log destinations created in or after 2013, which is the vast majority of them. Item 2 above - pre-2013 destinations when used with certain senders - is the exception.

**Comment by progrium** [comment]
@ben that's a great idea. I'd love a PR for that. If there is one that I've missed, please ping me from it.

**Comment by c-knowles** [comment]
Logspout also doesn't work out of the box for Loggly over TLS (fails with `!! x509: certificate signed by unknown authority`).

It looks like the go code is just defaulting to the OS level certs as @dan-compton mentioned above so in my case for Loggly I wrote this Dockerfile to grab their [certificate chain](https://www.loggly.com/docs/rsyslog-tls-configuration/) and install it:
```
FROM gliderlabs/logspout
RUN apk add -U --virtual build-dependencies curl ca-certificates \
  && curl https://logdog.loggly.com/media/logs-01.loggly.com_sha12.crt -o /usr/local/share/ca-certificates/loggly.crt \
  && update-ca-certificates \
  && apk del build-dependencies \
  && rm -rf /var/cache/apk/*
```

Logspout starts and connects ok over TLS now so hope this helps someone looking for the same thing. I don't have a papertrail account handy to check if the same solution works but assume it may do.

**Comment by cederberg** [comment]
Due to the logspout Dockerfile using `ONBUILD` the [solution above](https://github.com/gliderlabs/logspout/issues/147#issuecomment-239470682) doesn't work in an empty build context. Fails with the following:

```
Sending build context to Docker daemon  3.072kB
Step 1/4 : FROM gliderlabs/logspout
latest: Pulling from gliderlabs/logspout
cfc728c1c558: Pull complete 
97c60a365668: Pull complete 
f0c03c7cbd40: Pull complete 
Digest: sha256:9a807db8e9e484920c3228770da76b8af68e8704d8ca0965e68bf8cefc93cee1
Status: Downloaded newer image for gliderlabs/logspout:latest
# Executing 3 build triggers...
Step 1/1 : COPY ./build.sh /src/build.sh
lstat build.sh: no such file or directory
```

Fixed by building as follows:

```
$> touch build.sh modules.go
$> chmod +x build.sh
$> docker build -t logspout-loggly
```

Created a simple Docker image with the fix above + a simpler command-line. Probably won't be updated, though... https://hub.docker.com/r/qwayauser/logspout-loggly/

**Comment by gbolo** [comment]
@c-knowles @ben once this PR is merged #408 then you can mount additional CA certificate(s) and load them into logspout client TLS trust store with the following option: `LOGSPOUT_TLS_CA_CERTS`

more details of all possible options are in the PR

**Comment by gbolo** [comment]
fixed by #408
